parser/parser.go

```diff
@@ -91,13 +93,15 @@ func isIdempotentStmt(l *lexer, t token) (idempotent bool, err error) {
 	case tkUse:
 		return false, nil
 	case tkInsert:
-		return isIdempotentInsertStmt(l)
+		idempotent, t, err = isIdempotentInsertStmt(l)
 	case tkUpdate:
-		return isIdempotentUpdateStmt(l)
+		idempotent, t, err = isIdempotentUpdateStmt(l)
 	case tkDelete:
-		return isIdempotentDeleteStmt(l)
+		idempotent, t, err = isIdempotentDeleteStmt(l)
```

Comment:
Is there any case where the `t` returned by these 3 functions is different than `tkInvalid` when `err != nil`? If not, then we could remove it from those 3 functions because it doesn't really add anything of value.

Reply:
When inside a batch `t` could be the next DML starting token (`UPDATE`, `INSERT`, `DELETE`) or the end of the batch `APPLY`. If not inside a batch it should return `tkEOF` when no error or `tkInvalid` when an error happens (or not idempotent).